### PR TITLE
GEARPUMP-24, refactor DataSource API

### PR DESCRIPTION
**Pull request by manuzhang**

### CODE REVIEW

File: streaming/src/main/scala/io/gearpump/streaming/source/DataSource.scala [hunk 1]
```diff
@@ -29,10 +29,12 @@ import io.gearpump.{Message, TimeStamp}
...
  *  GenStringSource extends DataSource {
  *
- *    def open(context: TaskContext, startTime: Option[TimeStamp]): Unit = {}
+ *    def open(context: TaskContext, startTime: TimeStamp): Unit = {}
  *
- *    def read(batchSize: Int): List[Message] = {
- *      List.fill(batchSize)(Message("message"))
+ *    def hasNext: Boolean = true
+ *
+ *    def read(context: TaskContext): Unit = {
```
Comment (huafengw) [inline streaming/src/main/scala/io/gearpump/streaming/source/DataSource.scala]: The example code is not consistent with the API

File: external/kafka/src/main/scala/io/gearpump/streaming/kafka/KafkaSource.scala [hunk 2]
```diff
@@ -168,21 +169,11 @@ class KafkaSource(
...
-      fetch =>
-        var count = 0
-        while (count < batchSize) {
-          fetch.poll.flatMap(filterMessage).foreach(messageBuffer += _)
-          count += 1
-        }
-    }
-    messageBuffer.toList
+  override def read(): Message = {
+    fetchThread.flatMap(_.poll.flatMap(filterMessage)).orNull
```
Comment (huafengw) [inline external/kafka/src/main/scala/io/gearpump/streaming/kafka/KafkaSource.scala]: Why fetchThread has to be in an Option?

Comment (manuzhang) [inline external/kafka/src/main/scala/io/gearpump/streaming/kafka/KafkaSource.scala]: in case it's called uninitialized

**Comment by manuzhang** [comment]
any more comments on this PR ?

**Comment by whjiang** [comment]
generally +1. Do we have any other DataSource exist? E.g. the string collection one.If exist,we need to change all of them.

**Comment by manuzhang** [comment]
Do you mean the `CollectionDataSource` ? It's already updated.

**Comment by whjiang** [comment]
Do we have HBase source?

**Comment by manuzhang** [comment]
Nope. `KafkaSource` is the only external source in our codebase.

**Comment by whjiang** [comment]
Then +1.

**Comment by darionyaphet** [comment]
Hi @manuzhang  why change datasource API from `List[Message]` to `Message` ?

**Comment by manuzhang** [comment]
@darionyaphet please check https://issues.apache.org/jira/browse/GEARPUMP-24 for the motivation

**Pull request closed without merging**